File: 贪心/chapter3/3.2/No1842/main.cpp
#include <algorithm>
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
  string nextPalindrome(string num) {
    int n = num.size();
    if (n <= 1)
      return "";

    string half = num.substr(0, n / 2);
    string res = "";

    if (next_permutation(half.begin(), half.end())) {
      res = half;

      if (n % 2 == 1)
        res += num[n / 2];

      reverse(half.begin(), half.end());
      res += half;
      return res;
    }

    return "";
  }
};

class Solution2 {
public:
  string nextPalindrome(string num) {
    int n = num.size();
    if (n <= 1) {
      return "";
    }

    string half = num.substr(0, n / 2);
    string res = "";

    if (next_permutation(half.begin(), half.end())) {
      res = half;
      if (n & 1) {
        res += num[n / 2];
      }
      reverse(half.begin(), half.end());
      res += half;
      return res;
    }
    return res;
  }
};

signed main() {
  string num = "45544554";
  Solution2 sol;
  cout << sol.nextPalindrome(num) << endl;

  return 0;
}

File: 贪心/chapter3/3.2/No1842/main.py
class Solution:
    def nextPalindrome(self, num: str) -> str:
        n = len(num) // 2
        num_list = list(num)
        for i in range(n - 1, 0, -1):
            if num_list[i - 1] < num_list[i]:
                num_list[i], num_list[i - 1] = num_list[i - 1], num_list[i]
                num_list[n - i - 1], num_list[n - i] = (
                    num_list[n - i],
                    num_list[n - i - 1],
                )
                break
        ans = "".join(num_list)
        return ans if ans != num else ""


def main():
    num = "45544554"
    sol = Solution()
    print(sol.nextPalindrome(num))


main()
